src/userbase-server/purge.js

```diff
@@ -0,0 +1,461 @@
+import uuidv4 from 'uuid/v4'
+import logger from './logger'
+import setup from './setup'
+import connection from './connection'
+import connections from './ws'
+import userController from './user'
+import appController from './app'
+import adminController from './admin'
+import { getMsUntil1AmPst } from './utils'
+
+const MS_IN_AN_HOUR = 60 * 60 * 1000
+const MS_IN_A_DAY = 24 * MS_IN_AN_HOUR
+const TIME_TO_PURGE = 30 * MS_IN_A_DAY
+
+const ddbWhileLoop = async (params, ddbQuery, action) => {
+  let itemsResponse = await ddbQuery(params)
+  let items = itemsResponse.Items
+
+  await action(items)
+
+  // can be optimized with parallel scan
+  while (itemsResponse.LastEvaluatedKey) {
+    params.ExclusiveStartKey = itemsResponse.LastEvaluatedKey
+    itemsResponse = await ddbQuery(params)
+    items = itemsResponse.Items
+
+    await action(items)
+  }
+}
+
+const permanentDeleteDeletedItems = async (items, closeConnectedClients, permanentDelete) => {
+  const permanentDeletePromises = []
+
+  for (let i = 0; i < items.length; i++) {
+    const item = items[i]
+
+    if (item['deleted']) {
+      // closing connected clients tries to prevent lingering clients from inserting data while purge is under way
+      closeConnectedClients(item)
```

Comment:
Since each host runs the purge process, it should work nicely.

The only circumstance it won't work for is when an admin permanent deletes an app or user. In this case, the app or user gets placed in the permanent "deleted" table by the admin, rather than the purge process. So it won't even reach this area of the code base in that circumstance.

Reply:
But they won't run at the same time, no? So only one host will close the connections.

Reply:
When an item is deleted normally, there's a 30 day window before it will be placed in the "deleted" table. So each host will have a chance to close each item's connected clients.

Every night until the 30th day, the scan operation will scan the item's table, see the item is deleted, close all the item's connected clients, see the item has not been deleted for 30 days yet, and continue scanning.

So it's possible one host will close the connections at 1:15am and another host will close the connections at 2:33am.

But this way we can be highly confident when the purge process begins purging data, no clients will be connected (again, except in the case of permanent deletes).